### PR TITLE
BUG: DataFrame.append with timedelta64

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

AFAICT several of the existing tests are just wrong, xref #39122 cc @jorisvandenbossche 

Fixes (but havent yet added a test for) https://github.com/pandas-dev/pandas/issues/39037#issuecomment-756748197

### CODE REVIEW

File: pandas/core/dtypes/concat.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def _cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
     return arr.astype(dtype, copy=False)
 
 
-def concat_compat(to_concat, axis: int = 0):
+def concat_compat(to_concat, axis: int = 0, pretend_axis1: bool = False):
```
Comment (jreback) [inline pandas/core/dtypes/concat.py:100]: why don't you just allow axis=None and call it that case. this is very odd naming here (appreciate the de-duplication that this allows); so i guess a better question is this temporary?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
Can you summarize the behaviour that is changed / bug that is fixed?

File: pandas/core/dtypes/concat.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/concat.py:100]: > why don't you just allow axis=None and call it that case. this is very odd naming here

I want the naming to be very clear that this is a dont-try-this-at-home kludge (need to update the docstring to that effect)

> so i guess a better question is this temporary?

inasmuch as it wont be needed once we have 2D EAs, yes.

**Comment by jbrockmendel** [comment]
> Can you summarize the behaviour that is changed / bug that is fixed?

```
df = DataFrame({"a": pd.array([1], dtype="Int64")})
other = DataFrame({"a": [np.timedelta64("NaT", "ns")]})
result = df.append(other, ignore_index=True)

>>> result
      a
0     1
1  <NA>
>>> result.dtypes
a    Int64
dtype: object
```

```
df = DataFrame(columns=["a"]).astype("Int64")
other = DataFrame({"a": [np.timedelta64(1, "ns")]})

>>> result = df.append(other, ignore_index=True)
ValueError: Wrong number of dimensions. values.ndim != ndim [1 != 2]
```

File: pandas/core/dtypes/concat.py (hunk 1 above)
Comment (jreback) [reply pandas/core/dtypes/concat.py:100]: ok to the extent this is not removed anytime soon, can you come up with a better argument name or another way of doing this?

Comment (jbrockmendel) [reply pandas/core/dtypes/concat.py]: renamed + green

File: pandas/core/dtypes/concat.py [hunk 2]
```diff
@@ -72,6 +72,8 @@ def concat_compat(to_concat, axis: int = 0):
     ----------
     to_concat : array of arrays
     axis : axis to provide concatenation
+    ea_compat_axis : bool, default False
+        For ExtensionArray compat, behave as if axis == 1
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/concat.py:76]: Can you clarify what this means in practice to behave "as if axis=1"? Because I assume the arrays are still concatenated with axis=0?

File: pandas/tests/frame/methods/test_append.py [hunk 3]
```diff
@@ -165,23 +165,23 @@ def test_append_dtypes(self):
         df2 = DataFrame({"bar": np.nan}, index=range(1, 2))
         result = df1.append(df2)
         expected = DataFrame(
-            {"bar": Series([Timestamp("20130101"), np.nan], dtype="M8[ns]")}
+            {"bar": Series([Timestamp("20130101"), np.nan], dtype="object")}
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/methods/test_append.py:168]: Changing to ignore the all-NaN columns (although following the casting rules for common type of Timestamp and float) is still a breaking change, I would say. Certainly given that np.nan is generally used as missing value indicator (and allowed in other situations for datetime dtype).

File: pandas/tests/indexing/test_partial.py [hunk 4]
```diff
@@ -160,7 +160,8 @@ def test_partial_setting_mixed_dtype(self):
         df = DataFrame(columns=["A", "B"])
         df.loc[0] = Series(1, index=["B"])
 
-        exp = DataFrame([[np.nan, 1]], columns=["A", "B"], index=[0], dtype="float64")
+        # TODO: having this be float64 would not be unreasonable
+        exp = DataFrame([[np.nan, 1]], columns=["A", "B"], index=[0], dtype="object")
```
Comment (jorisvandenbossche) [inline pandas/tests/indexing/test_partial.py:164]: Why did this case change?

File: pandas/tests/reshape/concat/test_append.py [hunk 5]
```diff
@@ -334,29 +334,34 @@ def test_append_missing_column_proper_upcast(self, sort):
...
-        result_a = df.append(s, ignore_index=True)
+        result_a = df.append(ser, ignore_index=True)
         expected = DataFrame(
             [[np.nan, np.nan, 1.0, 2.0, date]], columns=["c", "d", "a", "b", "date"]
         )
         # These columns get cast to object after append
         expected["c"] = expected["c"].astype(object)
         expected["d"] = expected["d"].astype(object)
+        expected["date"] = expected["date"].astype(object)
+        # TODO: "date" might make sense to keep as dt64tz
```
Comment (jorisvandenbossche) [inline pandas/tests/reshape/concat/test_append.py:347]: This case should be possible to keep, I think? (independent from the value-dependent behaviour of all-NaN that I raised above)

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
I am personally a bit hesitant to change the all-NaN special case. That's longstanding behaviour (also for practical reasons), that will break code I think.

File: pandas/core/dtypes/concat.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/concat.py:76]: just matters for the dropping-empty check; will edit to clarify

File: pandas/tests/reshape/concat/test_append.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/reshape/concat/test_append.py:347]: it should be possible, but i havent found a straightforward way to make it happen.  i was thinking your ArrayProxy might be useful for this

**Comment by jbrockmendel** [comment]
@jorisvandenbossche helpful comments, thank you.  i found a nice way to retain the None/nan behavior while fixing the original bug: supplementing JoinUnit.is_na with a check that it is a compatible NA.

File: pandas/core/internals/concat.py [hunk 6]
```diff
@@ -233,6 +227,24 @@ def dtype(self):
...
+        if not self.is_na:
+            return False
+        if self.block is None:
+            return True
+
+        if self.dtype == object:
+            values = self.block.values
+            return all(
+                is_valid_nat_for_dtype(x, dtype) for x in values.ravel(order="K")
+            )
```
Comment (jorisvandenbossche) [inline pandas/core/internals/concat.py:243]: This is not only required for object dtype, I think. Also float NaN is considered "all NaN" when it comes to ignoring the dtype in concatting dataframes:

```
In [39]: pd.concat([pd.DataFrame({'a': [np.nan, np.nan]}), pd.DataFrame({'a': [pd.Timestamp("2012-01-01")]})])
Out[39]: 
           a
0        NaT
1        NaT
0 2012-01-01

```

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/core/internals/concat.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/internals/concat.py:243]: the non-object case is handled below on L245-246.  or do you have something else in mind?

Comment (jorisvandenbossche) [reply pandas/core/internals/concat.py:243]: Does my snippet above work with this PR? 
(if so, then I don't fully understand why the changes to `test_append_empty_frame_to_series_with_dateutil_tz` are needed)

Comment (jbrockmendel) [reply pandas/core/internals/concat.py:243]: > Does my snippet above work with this PR?

yes it does

> (if so, then I don't fully understand why the changes to test_append_empty_frame_to_series_with_dateutil_tz are needed)

I think that's driven by something sketchy-looking in get_reindexed_values, will see if that can be addressed.

Comment (jbrockmendel) [reply pandas/core/internals/concat.py]: better?

**Comment by jbrockmendel** [comment]
@jorisvandenbossche i think your comments have all been addressed, lmk if i missed anything

**Comment by jbrockmendel** [comment]
@jorisvandenbossche gentle ping before your day ends

**Comment by jorisvandenbossche** [comment]
Can you add a whatsnew?

**Comment by jbrockmendel** [comment]
will do, thanks

File: pandas/core/internals/concat.py [hunk 7]
```diff
@@ -232,6 +227,28 @@ def dtype(self):
...
+            return True
+
+        if self.dtype == object:
+            values = self.block.values
+            return all(is_valid_na_for_dtype(x, dtype) for x in values.ravel(order="K"))
+
+        if self.dtype.kind == dtype.kind == "M" and not is_dtype_equal(
+            self.dtype, dtype
+        ):
+            # fill_values match but we should not case self.block.values to dtype
```
Comment (jorisvandenbossche) [inline pandas/core/internals/concat.py:246]:
```suggestion
            # fill_values match but we should not cast self.block.values to dtype
```

File: pandas/core/internals/concat.py [hunk 8]
```diff
@@ -232,6 +227,28 @@ def dtype(self):
             return blk.dtype
         return ensure_dtype_can_hold_na(blk.dtype)
 
+    def is_valid_na_for(self, dtype: DtypeObj) -> bool:
+        """
+        Check that we are all-NA of a type/dtype that is compatible with this dtype.
```
Comment (jorisvandenbossche) [inline pandas/core/internals/concat.py:232]:
```suggestion
        Check that we are all-NA of a type/dtype that is compatible with this dtype.
        Augments `self.is_na` with an additional check of the type of NA values.
```

That's correct? 

(I was missing before that this basically is `self.is_na` but a bit stricter for a few cases)

Comment (jbrockmendel) [reply pandas/core/internals/concat.py:232]: yes

File: pandas/tests/reshape/concat/test_append.py [hunk 9]
```diff
@@ -350,13 +350,14 @@ def test_append_empty_frame_to_series_with_dateutil_tz(self):
...
-
-        result_b = result_a.append(s, ignore_index=True)
+        result_b = result_a.append(ser, ignore_index=True)
         tm.assert_frame_equal(result_b, expected)
 
         # column order is different
         expected = expected[["c", "d", "date", "a", "b"]]
-        result = df.append([s, s], ignore_index=True)
+        dtype = Series([date]).dtype
+        expected["date"] = expected["date"].astype(dtype)
```
Comment (jorisvandenbossche) [inline pandas/tests/reshape/concat/test_append.py:359]: Is this still needed? (might be a left-over from astyping it to object before)

**Review by jorisvandenbossche — APPROVED** [review]
Last minor comment at https://github.com/pandas-dev/pandas/pull/39574/files#r575450869

Nice clean-up!

File: pandas/tests/reshape/concat/test_append.py (hunk 9 above)
Comment (jbrockmendel) [reply pandas/tests/reshape/concat/test_append.py:359]: you're right, updated

**Review by jreback — APPROVED** [review]
